File: packages/tensor-factory/src/tensor_factory/codec.py
"""Quantize a normalized box to four 8-bit values -- the detection contract.

The model output is deliberately tiny: a helicoil box is four ``uint8`` values, one
per normalized coordinate. At 480 px the quantization step is ``480 / 255 ~= 1.88``
px, so the worst-case round-trip error is under one pixel -- comfortably inside the
3 px budget -- while keeping the model head and post-processing in 8-bit math.
"""

from __future__ import annotations

from collections.abc import Sequence

from .geometry import BBox

_MAX = 255  # 256 levels, 0..255


def _q(v: float) -> int:
    q = round(v * _MAX)
    return 0 if q < 0 else _MAX if q > _MAX else q


def encode_uint8(box: BBox) -> tuple[int, int, int, int]:
    """Encode a normalized box to four ``uint8`` values ``(x1, y1, x2, y2)``."""
    return (_q(box.x1), _q(box.y1), _q(box.x2), _q(box.y2))


def decode_uint8(values: Sequence[int]) -> BBox:
    """Decode four ``uint8`` values back to a normalized :class:`BBox`.

    Order is repaired and coordinates clamped, so a model that emits ``x2 < x1`` for
    a near-degenerate box still yields a valid box rather than raising.
    """
    if len(values) != 4:
        raise ValueError(f"expected 4 values, got {len(values)}")
    for v in values:
        if not 0 <= v <= _MAX:
            raise ValueError(f"value {v!r} outside uint8 range [0, {_MAX}]")
    x1, y1, x2, y2 = (v / _MAX for v in values)
    return BBox.clamped(x1, y1, x2, y2)


def max_error_px(image_size: int) -> float:
    """Worst-case per-coordinate round-trip error, in pixels, at ``image_size``."""
    return image_size / _MAX / 2.0

File: studio/js/codec.js
// 4×uint8 detection codec + BBox helpers — browser port of tensor_factory.codec
// / tensor_factory.geometry. A box is normalized xyxy in [0,1], top-left origin.
// Encoding each coord to one uint8 gives a ~480/255 ≈ 1.88 px step at 480 px,
// keeping round-trip error under 1 px — the repo's on-the-wire contract.

export function clamp01(v) {
  return v < 0 ? 0 : v > 1 ? 1 : v;
}

/** Order + clamp a raw box into a valid normalized xyxy box. */
export function orderClamp({ x1, y1, x2, y2 }) {
  x1 = clamp01(x1);
  y1 = clamp01(y1);
  x2 = clamp01(x2);
  y2 = clamp01(y2);
  if (x2 < x1) [x1, x2] = [x2, x1];
  if (y2 < y1) [y1, y2] = [y2, y1];
  return { x1, y1, x2, y2 };
}

/** Normalized box → [x1,y1,x2,y2] as four uint8 (0–255). */
export function encodeUint8(box) {
  const b = orderClamp(box);
  return [b.x1, b.y1, b.x2, b.y2].map((v) => Math.round(clamp01(v) * 255));
}

/** Four uint8 → normalized box, repairing order + clamping. */
export function decodeUint8(values) {
  const [x1, y1, x2, y2] = values.map((v) => Math.min(255, Math.max(0, v)) / 255);
  return orderClamp({ x1, y1, x2, y2 });
}

/** Worst-case per-coordinate round-trip error in pixels. */
export function maxErrorPx(imageSize) {
  return imageSize / 255 / 2;
}

export function area(box) {
  return Math.max(0, box.x2 - box.x1) * Math.max(0, box.y2 - box.y1);
}

/** Normalized box → COCO [x, y, w, h] in absolute pixels. */
export function toCocoBbox(box, width, height) {
  const b = orderClamp(box);
  return [b.x1 * width, b.y1 * height, (b.x2 - b.x1) * width, (b.y2 - b.y1) * height];
}
